bin/spice/cmd/upgrade.go

```diff
@@ -45,103 +48,143 @@ spice upgrade
+		if models {
+			flavor = constants.FlavorAI
 		}
 
-		tmpDirName := strconv.FormatInt(time.Now().Unix(), 16)
-		tmpDir := filepath.Join(spiceBinDir, tmpDirName)
-
-		err = os.Mkdir(tmpDir, stat.Mode())
+		release, err := github.GetLatestRuntimeRelease()
 		if err != nil {
-			slog.Error("upgrading the spice binary", "error", err)
-			return
+			slog.Error("installing runtime", "error", err)
+			os.Exit(1)
 		}
-		defer os.RemoveAll(tmpDir)
 
-		err = github.DownloadAsset(release, tmpDir, assetName)
+		err = rtcontext.InstallOrUpgradeRuntime(flavor, accelerated) // retain the current accelerator setting for upgrades
 		if err != nil {
-			slog.Error("downloading the spice binary", "error", err)
-			return
+			slog.Error("installing runtime", "error", err)
+			os.Exit(1)
 		}
 
-		tempFilePath := filepath.Join(tmpDir, constants.SpiceCliFilename)
+		slog.Info(fmt.Sprintf("Spice runtime upgraded to %s successfully.", release.TagName))
+	},
+}
 
-		err = util.MakeFileExecutable(tempFilePath)
+func upgradeCli(force bool, rtcontext *context.RuntimeContext) {
+	slog.Info("Checking for latest Spice CLI release...")
+	release, err := github.GetLatestCliRelease()
+	if err != nil {
+		slog.Error("checking for latest release", "error", err)
+		return
+	}
+
+	cliVersion := version.Version()
+
+	runtimeUpgradeRequired, err := rtcontext.IsRuntimeUpgradeAvailable()
+	if err != nil {
+		slog.Error("checking for runtime upgrade", "error", err)
+		return
+	}
+
+	if cliVersion == release.TagName && runtimeUpgradeRequired != "" && !force {
+		slog.Info(fmt.Sprintf("Using the latest version %s. No upgrade required.", release.TagName))
+		return
+	}
+
+	assetName := github.GetAssetName(constants.SpiceCliFilename)
+	spiceBinDir := filepath.Join(rtcontext.SpiceRuntimeDir(), "bin")
+
+	slog.Info("Upgrading the Spice.ai CLI ...")
+
+	stat, err := os.Stat(spiceBinDir)
+	if err != nil {
+		slog.Error("upgrading the spice binary", "error", err)
+		return
+	}
+
+	tmpDirName := strconv.FormatInt(time.Now().Unix(), 16)
+	tmpDir := filepath.Join(spiceBinDir, tmpDirName)
+
+	err = os.Mkdir(tmpDir, stat.Mode())
+	if err != nil {
+		slog.Error("upgrading the spice binary", "error", err)
+		return
+	}
+	defer os.RemoveAll(tmpDir)
+
+	err = github.DownloadAsset(release, tmpDir, assetName)
+	if err != nil {
+		slog.Error("downloading the spice binary", "error", err)
+		return
+	}
+
+	tempFilePath := filepath.Join(tmpDir, constants.SpiceCliFilename)
+
+	err = util.MakeFileExecutable(tempFilePath)
+	if err != nil {
+		slog.Error("upgrading the spice binary", "error", err)
+		return
+	}
+
+	releaseFilePath := filepath.Join(spiceBinDir, constants.SpiceCliFilename)
+
+	// On Windows, it is not possible to overwrite a binary while it's running.
+	// However, it can be moved/renamed making it possible to save new release with the original name.
+	if util.IsWindows() {
+		runningCliTempLocation := filepath.Join(spiceBinDir, constants.SpiceCliFilename+".bak")
```

Comment:
Don't call this .bak. Copy the binary to the Window temporary directory, under a temp path, and run it from there. That way it will get cleaned up. Also, clean it up on exit.